src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -185,12 +188,56 @@ private static uint HandleEventShutdownComplete(State state, ref ConnectionEvent
             return MsQuicStatusCodes.Success;
         }
 
+        public void RemoveStream(MsQuicStream stream)
+        {
+            lock (_state)
+            {
+                _state.StreamCount--;
+            }
+
+            if (_state.Connection == null && _state.StreamCount == 0)
```

Comment:
I remember that `_state.Connection` initially was not always set, because `_state` is specifically rooted in GC. Now if I understood correctly, with your change it will be always set, so even if connection is thrown away and forgotten, it will be also kept alive by GC. Streams are not specifically rooted, so even if they are forgotten, they will be able to be CG'ed and their finalizer will call RemoveStream which will eventually reduce stream count to 0. But `_state.Connection` is set to null only on Connection's dispose/finalize, first was not called because connection was thrown away, second was not called because it's still alive from GC's perspective. So in this situation it seems that we will not go inside this `if`, `_stateHandle` will never be freed and connection will be alive forever. Am I missing something?

Reply:
yes, I think you are right that the "forgotten" e.g. not disposed case would be problematic. 
I could rework the code to set _state.Connection only when there are active streams to make the connection disposable by GC. (as this change alone is really focused only on avoiding the access violation) 

Alternatively we can link just the State like you did in #52776 but I really did not like exposing the inner guts outside of the class. I think it is crucial to reach consensus on directions and possibly look at what else we need besides not crashing.